### PR TITLE
Adding tests for elasticsearch version request retries

### PR DESCRIPTION
**Pull request by lancewf**
Signed-off-by: Lance Finfrock <lfinfrock@chef.io>

Adding test for ticket #1307

### CODE REVIEW

File: omnibus/files/private-chef-cookbooks/private-chef/spec/libraries/helper_spec.rb [hunk 1]
```diff
@@ -114,6 +114,64 @@
     end
 
     context 'when elastic search is unavailable' do
+      it 'should return 2 but the request fails first time only' do
```
Comment (marcparadise) [inline omnibus/files/private-chef-cookbooks/private-chef/spec/libraries/helper_spec.rb]: The tests look good, but I'm thinking they could be a shared example that exists locally this context, and uses `let` or similar  to set number of failures?

File: omnibus/files/private-chef-cookbooks/private-chef/spec/libraries/helper_spec.rb [hunk 2]
```diff
@@ -114,6 +114,64 @@
...
+        helper = described_class.new(node)
+        @times_called = 0
```
Comment (phiggins) [inline omnibus/files/private-chef-cookbooks/private-chef/spec/libraries/helper_spec.rb]: Is `@times_called` used anywhere?

File: omnibus/files/private-chef-cookbooks/private-chef/spec/libraries/helper_spec.rb [hunk 3]
```diff
@@ -114,6 +114,64 @@
...
+        allow(Chef::HTTP).to receive(:new).with(external_url).and_return(client)
+
+        expect(client).to receive(:get).with('').exactly(3).times.and_raise('Bad connection')
+        expect(client).to receive(:get).with('').and_return(response)
+
+        allow(helper).to receive(:sleep).and_return(0)
+        expect(helper.elastic_search_major_version).to eq(2)
+      end
```
Comment (phiggins) [inline omnibus/files/private-chef-cookbooks/private-chef/spec/libraries/helper_spec.rb]: It looks like there are two copies of this test?

File: omnibus/files/private-chef-cookbooks/private-chef/spec/libraries/helper_spec.rb (hunk 1 above)
Comment (phiggins) [inline omnibus/files/private-chef-cookbooks/private-chef/spec/libraries/helper_spec.rb]: You could also use a loop for these tests that only differ by the count:

```
1.upto(4) do |n|
  it "eventually returns 2 after #{n} failures" do
    # ...
```

Comment (marcparadise) [inline omnibus/files/private-chef-cookbooks/private-chef/spec/libraries/helper_spec.rb]: That would also work, though lately I have a preference for shared context in the same file vs dynamically created tests, for ease of understandability and to keep the amount of test-related "logic" to a minimum.

Comment (phiggins) [inline omnibus/files/private-chef-cookbooks/private-chef/spec/libraries/helper_spec.rb]: Shared context has always seemed like wonky rspec voodoo to me, so I think it's just a matter of taste. :)

File: omnibus/files/private-chef-cookbooks/private-chef/spec/libraries/helper_spec.rb (hunk 2 above)
Comment (lancewf) [inline omnibus/files/private-chef-cookbooks/private-chef/spec/libraries/helper_spec.rb]: No, `@times_called` is not used. I will remove it. Good catch.

File: omnibus/files/private-chef-cookbooks/private-chef/spec/libraries/helper_spec.rb (hunk 3 above)
Comment (lancewf) [inline omnibus/files/private-chef-cookbooks/private-chef/spec/libraries/helper_spec.rb]: There is two. Removing the extra test.

**Comment by lancewf** [comment]
I switched the code to use shared context and also removed the not used variable.

File: omnibus/files/private-chef-cookbooks/private-chef/spec/libraries/helper_spec.rb [hunk 4]
```diff
@@ -113,7 +113,40 @@
       end
     end
 
+    shared_context "elastic search version request fails" do
```
Comment (phiggins) [inline omnibus/files/private-chef-cookbooks/private-chef/spec/libraries/helper_spec.rb]: Does `shared_context` let you share specs? I thought it was only for sharing test setup.

Comment (phiggins) [inline omnibus/files/private-chef-cookbooks/private-chef/spec/libraries/helper_spec.rb]: As long as the tests are actually running this seems good. :)

Comment (lancewf) [inline omnibus/files/private-chef-cookbooks/private-chef/spec/libraries/helper_spec.rb]: I copied the idea from [here](https://github.com/chef/chef-server/blob/608dbe94d15822a31849952e13549744fc40a702/oc-chef-pedant/spec/api/versioned_behaviors/server_api_v1_spec.rb#L90). 

I did ensure that the test is working as intended by changing the `times` variable to 7 and watching it fail.

**Pull request merged**